Core/SAI: Remove no longer used event types & params

offl opened the pull request:
**Changes proposed:**

-  remove SMART_EVENT_TARGET_HEALTH_PCT, reasons here 41275c056e7a03d140e36924b42f043625fe1a0d
-  remove SMART_EVENT_FRIENDLY_HEALTH, no longer used since 2020
-  remove SMART_EVENT_IS_BEHIND_TARGET, no longer used since 2020
-  remove params of SMART_EVENT_RESPAWN (SMART_SCRIPT_RESPAWN_CONDITION), (0c3d11c71a056acdd392ddc219185444f63b6372), we already have `conditions`

**Target branch(es):** 

3.3.5

**Issues addressed:**

none

**Tests performed:**

didn't checked

**Known issues and TODO list:**

none, so far

offl committed:
Core/SAI: Remove no longer used event types & params

jackpoz commented on src/server/game/AI/SmartScripts/SmartScript.cpp:
```diff
@@ -3136,10 +3098,7 @@ void SmartScript::ProcessEvent(SmartScriptHolder& e, Unit* unit, uint32 var0, ui
         {
             if (!GetBaseObject())
```
this might not be needed anymore as there are no more checks to do with GetBaseObject()

offl committed:
Move SMART_EVENT_RESPAWN to 'no params'

ccrs commented:
not a fan, I see no gain from removing those events simply because they are not used right now.

about the SMART_EVENT_RESPAWN, I see it reasonable, nothing to complain there

jackpoz commented:
that's a fair point, even if we don't use some of those SAI events/actions it doesn't hurt to leave them there (at least the ones that don't have another implementation already)

offl commented:
I was a little tired and didn't explained why I want to remove them. They are removed not because they're not used but because I know that they should not be used

I spent a lot of time to convert all SMART_EVENT_FRIENDLY_HEALTH to PTC and don't want to let people continue to use it for absolutely no reason. That event was always really bad because it's simply dumb to use HP value instead of HP percentage. Especially because we have different dungeon modes with different HP values and there are no signs that Blizz checks HP value and not percentages

Regarding SMART_EVENT_TARGET_HEALTH_PCT and SMART_EVENT_IS_BEHIND_TARGET, well, I have to say that I became a little obsessed recently with Blizz's original AI actions, hooks and stuff like that. I didn't found such AI hooks. I'm aware it does not mean that they don't have them.

It may be risky to remove those event types, especially SMART_EVENT_TARGET_HEALTH_PCT, I'm aware of it. We can restore them later if they will be needed. However at this point I don't see why we should leave posibility to use them. Especially because they are useless, I mean core does not care when WE ARE TRYING TO use those spells(both events are used only for spells), it already checks attributes(SMART_EVENT_IS_BEHIND_TARGET) and aura states(SMART_EVENT_TARGET_HEALTH_PCT) of spells used in those events and does not care when you want you cast those spells. By using those event types we check everything twice. Also I prefer to remove things completely rather than wait to people to use them again in scripts for no reason because converting them again takes time

***

>events are tools to be able to script stuff, why having less tools?

I have completely different opinion for that now. Some people say 'SAI is too much limited'. That is not entirely correct. I mean there indeed should be more things but at the same time there should be less things or some things should be handled completely differently or should be strongly limited and not only for DB(#25764). Speaking of safety, some parts of SAI are old ticking bombs that explodes periodically. Solving one problem takes lot of time and breaks tens of scripts(#26728) and that will happen over and over again

We have all Blizz's AI actions, hooks and even params of their analogue of SAI for years(actually from beginning, from alpha). After checking that stuff, including leaked recently enums, I ended up with such conclusion:

Some of their flags has comments like 'TALK TO A PROGRAMMER FIRST'. This is my favorite so far:
>Always, ALWAYS tries to stand right on top of his move to target. ALWAYS!!	

Some things are limited to programmers only
>(Programmer Only) Don't consume charge if item deleted

So their development team has programmers and designers. Because of all those comments, I think their designers probably don't know anything about how core works(same with 90%+ of emu DB devs). Designers uses analogue of our SAI. Do they script bosses? Maybe only if they know how to do it, I mean if they not create only quests, creatures and stuff like that? Maybe there is another type of designers or maybe they work in team and everybody does only what they know? Maybe only programmers script bosses? Unfortunately I don't know much about all that stuff but unexpectantly here it really does matter to know how development process of a game works

At the same time their analogue of SAI looks to me as system which is used not only for simple scripts but even for bosses because of the way that system constructed. It affects all creatures and objects in game, not only the ones we script only in DB. It doesn't look like they want to have duplicating logic and avoid to use that system for bosses. But at the same time it doesn't look like their analogue of SAI is used to script everything we script by using BossAI because it's too limited

Designers can do in scripts only limited amount of things(so do we, DB devs can operate only with limited amount of actions and event types), as it's supposed to be. But even that way Blizz's devs somehow can make really weird things in their DB scripts(so do we)

>We have all Blizz's AI actions, hooks and even their params of their analogue of SAI for years.

Would be nice to focus on that when creating something new or adjusting something old. Simply to not create monsters like #26728. The only exceptions are iternal systems we don't even have implemented, in other words in some cases we have no chose and have to create and use event types and AI actions which Blizz never used. But in lots of other cases I simply see no reason to create something different that will just cause problems in future

jackpoz commented:
SMART_EVENT_TARGET_HEALTH_PCT and SMART_EVENT_IS_BEHIND_TARGET are not bound to cast spells, one could even just send a chat message when those events happen for whatever reason.
SMART_EVENT_FRIENDLY_HEALTH has been the only working health related event for a long time until SMART_EVENT_TARGET_HEALTH_PCT was implemented properly.

I would prefer to keep these 3 events and remove the rest. We don't have to use them in TC, we don't have to approve fixes by users that include them, but it costs very little to keep them implemented.

offl commented:
>SMART_EVENT_TARGET_HEALTH_PCT and SMART_EVENT_IS_BEHIND_TARGET are not bound to cast spells, one could even just send a chat message when those events happen for whatever reason.

Should never happen. This is not even a thing

>We don't have to use them in TC, we don't have to approve fixes by users that include them, but it costs very little to keep them implemented.

#20648 dude literally changed WDB fields because creatures had too less mana to cast spells from his scripts
That was pushed to the source, nobody checked that, nobody paid attention
#24245 too nobody checked. Dude used that flag in nearly all his scripts for months. Nobody paid attention, all that stuff was added to the source. How I removed it knowing it's needed in some cases? Well, I removed it from DB and then spawned all those creatures and checked if they still can cast spell without that flag. Had a damn good time
#26728 why is this even a problem? Why people made 500 creatures passive or defensive? What was on their mind? Of course in lots of cases they just used default value(0) but I can swear some people used 0 or 1 because they were thinking they have to do it. Or they will overwrite default react state(faction, civilian flag) which is not even possible. I checked those 500 creatures and I know what I'm talking about
What I think about SAI's scripting library? Can't even chose what to say. Let's say hundreds of scripts are so awful that they requires no updates but full rewrite. Thousands of scripts should updated with modern standards

Meh, 2 hours of writing for nothing
No regrets, now I know what that Blizz's SAI is and who uses it, sadly you don't understand that

ccrs commented:
Its nice to see how you raise points, answer them yourself and then extract your own assuptions based on those autoanswered points about the perception you have on people's knowledge and awareness

jackpoz commented:
After talking about the issues that these events can cause, especially when used incorrectly, and how long it takes for us to fix those issues, it looks important enough for us to not support these events anymore.

jackpoz commented:
One thing that really makes no sense with these events is to use them to replicate the checks of Spell::CheckCast:
- take an example with a spell that requires 1k mana, target with hp < 20% and caster behind target
- which event would you pick ? SAI_EVENT_TARGET_HP_PCT ? SAI_EVENT_TARGET_BEHIND ? SAI_EVENT_HAS_MANA ?
- whatever is picked, it will not be enough to apply ALL the checks that Spell::CheckCast does, making it an half-baked solution where the event is triggered but the action fails

What would make sense in this would be a SAI_EVENT_CAN_CAST that does all the checks that Spell::CheckCast() does, but then that's the same as just casting the spell and checking the result.

@offl thanks for helping me realize how non-sensical using these events as "spell checks" is

jackpoz commented:
I will work on implementing a warning on SAI actions/events that we suggest not to use. It will allow TC to get notified if a contribution uses them and at the same time allow users to keep using them if they want. Being a log, it can be turned off through configs.

offl commented:
>take an example with a spell that requires 1k mana, target with hp < 20% and caster behind target

There is no such spells, SMART_EVENT_TARGET_MANA_PCT is not even used currently(surprising)
